packages/execution-environments/src/common/BaseSnapExecutor.ts

```diff
@@ -123,28 +124,39 @@ export class BaseSnapExecutor {
           handler !== undefined,
           `No ${handlerName} handler exported for snap "${target}`,
         );
-        // TODO: fix type
-        return this.executeInSnapContext(target, () => handler(args as any));
+        // TODO: fix handler args type cast
+        const result = await this.executeInSnapContext(target, () =>
+          handler(args as any),
+        );
+        assert(
+          isValidJson(result),
+          new TypeError('Received non JSON serializable value'),
+        );
+        return result;
       },
       this.onTerminate.bind(this),
     );
   }
 
-  private errorHandler(error: unknown, data: Record<string, unknown>) {
+  private errorHandler(error: unknown, data: Record<string, Json>) {
     const constructedError = constructError(error);
     const serializedError = serializeError(constructedError, {
       fallbackError,
       shouldIncludeStack: false,
     });
+
+    // We're setting it this way to avoid sentData.stack = undefined
+    const sentData: Json = { ...data };
+    if (constructedError?.stack) {
+      sentData.stack = constructedError.stack;
+    }
```

Comment:
```suggestion
    // We're setting it this way to avoid sentData.stack = undefined
    const sentData: Json = { ...data, stack: constructedError?.stack ?? null };
```